proxy/src/main/java/com/velocitypowered/proxy/plugin/VelocityPluginManager.java

```diff
@@ -0,0 +1,112 @@
+package com.velocitypowered.proxy.plugin;
+
+import com.velocitypowered.api.plugin.PluginCandidate;
+import com.velocitypowered.api.plugin.PluginContainer;
+import com.velocitypowered.api.plugin.PluginManager;
+import com.velocitypowered.api.plugin.meta.PluginDependency;
+import com.velocitypowered.proxy.VelocityServer;
+import com.velocitypowered.proxy.plugin.loader.JavaPluginLoader;
+import com.velocitypowered.proxy.plugin.util.PluginDependencyUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.io.IOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.*;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class VelocityPluginManager implements PluginManager {
+    private static final Logger logger = LogManager.getLogger(VelocityPluginManager.class);
+
+    private final Map<String, PluginContainer> plugins = new HashMap<>();
+    private final Map<Object, PluginContainer> pluginInstances = new HashMap<>();
```

Comment:
IMO, this should be an `IdentityHashMap`. Otherwise it's possible to mess this up pretty easily by implementing a weird `equals` method in the plugin main class. There should never be more than once instance of the plugin main class, so the identity check should do the job.